iceoryx_hoofs/include/iceoryx_hoofs/internal/design_pattern/polymorphic_handler.inl

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_DESIGN_PATTERN_POLYMORPHIC_HANDLER_INL
+#define IOX_HOOFS_DESIGN_PATTERN_POLYMORPHIC_HANDLER_INL
+
+#include "iceoryx_hoofs/design_pattern/polymorphic_handler.hpp"
+#include "iceoryx_hoofs/design_pattern/static_lifetime_guard.hpp"
+#include <atomic>
+#include <type_traits>
+
+namespace iox
+{
+namespace design_pattern
+{
+
+namespace detail
+{
+
+template <typename Interface>
+void DefaultHooks<Interface>::onSetAfterFinalize(Interface&, Interface&) noexcept
+{
+    // we should not use an error handling construct (e.g. some IOX_ASSERT) here for dependency reasons
+    // we could in principle do nothing by default as well, but the misuse failure should have visible consequences
+    std::terminate();
+}
+
+} // namespace detail
+
+// The get method is considered to be on the hot path and works as follows:
+//
+// On first call (in a thread):
+// 1. localHandler is initialized
+//    - getCurrent is called
+//    - instantiates singleton instance()
+//    - instantiates default handler
+//    - sets m_current of instance to default instance (release)
+
+// 2. If any thread changes the handler with set or reset, it will:
+//    - be detected by comparison with local handler
+//    - set the current handler to the new handler
+//
+
+// On any call after the handler was changed in another thread
+// 1. We detect the change by comparison with the localHandler (can assume that addresses are unique by
+// design as we use singletons in static memory)
+// 2. If it was changed we perform a synchronizing load and obtain the current handler
+// Note that it may change again but this is unavoidable if it can change concurrently
+// without locks and not a problem.
+template <typename Interface, typename Default, typename Hooks>
+Interface& PolymorphicHandler<Interface, Default, Hooks>::get() noexcept
+{
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables) false positive, thread_local
+    thread_local Interface* localHandler = getCurrent(); // initialized once per thread on first call
+
+    // weak load, required since it could be set in another thread concurrently
+    auto* currentHandler = instance().m_current.load(std::memory_order_relaxed);
```

Comment:
How about having a `getCurrentSync` and a `getCurrentRelaxed` method? This makes it easier to understand what's going on without looking at the `getCurrent` implementation.

Reply:
Yes, it is more expressive that way.